stm32g0: add fdcan2 in dts file

adri1mart1 opened the pull request:
Hello,

Following my recent [opened discussion](https://github.com/zephyrproject-rtos/zephyr/discussions/67444), I have tried to deal with the dual can controller of the stm32g0.

Right now, it is not supported, however, it seems very small changes are needed to make it work.

STM32G0 shares the same interrupt line for both its CAN controllers and about a year ago, the shared interrupt line had been added (see [shared interrupt PR](https://github.com/zephyrproject-rtos/zephyr/pull/61422))

Another interesting information is about the driver itself, the can_stm32_fdcan.c implements a macro to generate fdcan instances based on all enabled fdcan controller nodes in dts that are with a status okay (see [macro](https://github.com/zephyrproject-rtos/zephyr/blob/04de43b1a119665532b6b2a0c6bbcbefd47eb7d6/drivers/can/can_stm32_fdcan.c#L689)).

So on my side, i've just added the fdcan2 for the stm32g0 in the stm32g0b1.dtsi and also on the nucleo_g0b1re board with appropriate pinout. 

To test my setup, I have used the nucleo_g0b1re board with 2 external transceivers on a breadboard to set up a communication between fdcan1 and fdcan2. I used a PEAK CAN probe to check frames are well physically transferred from one to another.
I have tested only fdcan1, then only fdcan2, then both, it seems to work great !

Right now, there is no sample program from zephyr ready to go with this configuration so I had to update the counter example for my needs and I'm definitely not sure it is absolutely needed.

Important note on this modification. In the nucleo_g0b1re dts file, I have added the fdcan2 node with the correct pinout. However, fdcan2 is disabled by default whereas fdcan1 is enabled. 

If you enable fdcan1 and fdcan2, you will get this error:

```
[151/155] Generating isr_tables.c
FAILED: zephyr/isr_tables.c .../build/zephyr/isr_tables.c 
[...]
gen_isr_tables.py: error: multiple registrations at table_index 22 for irq 22 (0x16)
Existing handler 0x8009c8b, new handler 0x8009c8b
Has IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?
```

You must add the `CONFIG_SHARED_INTERRUPT=y` flag to make it work. This could be confusing for anyone who is not familiar with the fact that stm32g0 fdcan controller is concerned about shared interrupt. 

I am not sure if the zephyr community wants to live with it, improve the comment above when the error is raised, add a specific paragraph in the nucleo_g0b1re doc page. Any feedback will be very appreciated.

Regards,
Adrien M.

Commit:
soc: stm32g0: add fdcan2

The STM32G0 soc has 2 CAN controllers. The 2nd on was not working
with zephyr yet as both controllers shares the same IRQ. Recently, the
shared irq system was integrated on now, both can controllers can work
on this chip. Shared interrupts must be enabled only if both can
controllers are enabled.

Signed-off-by: Adrien MARTIN <adrienmar@kickmaker.net>

henrikbrixandersen reviewed (changes requested):
@erwan Do you know if all `compatible = "st,stm32-fdcan"` instances across the STM32 use shared interrupts?

If so, we need to automatically enable shared IRQs on the driver Kconfig level. If not, we need to enable it on SoC Kconfig level. Regardless of which it is, we'll likely need a new Kconfig function for checking the number of status=okay nodes with a given compatible.

erwango commented:
@henrikbrixandersen From what I can see, only stm32g0b1 and stm32g0c1 provide FDCAN, but both would require shared interrupts.

henrikbrixandersen commented:
> @henrikbrixandersen From what I can see, only stm32g0b1 and stm32g0c1 provide FDCAN, but both would require shared interrupts.

Thanks. Based on that, I'd like to see something like the following added to this PR:

```
diff --git a/soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx b/soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx
index 86c47b158e..05d2116225 100644
--- a/soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx
+++ b/soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx
@@ -11,4 +11,7 @@ config SOC
 config NUM_IRQS
        default 31
 
+config SHARED_INTERRUPTS
+       default y if $(dt_nodelabel_enabled,fdcan1) && $(dt_nodelabel_enabled,fdcan2)
+
 endif # SOC_STM32G0B1XX
```

adri1mart1 commented on boards/arm/nucleo_g0b1re/nucleo_g0b1re.dts:
```diff
@@ -187,6 +187,16 @@ zephyr_udc0: &usb {
+	bus-speed-data = <1000000>;
+	status = "disabled";
+};
+
```
Adding this fdcan2 was obviously to use both CAN controllers on this nucleo board. 
Keeping it disabled was not to break any compilation as explained in my initial comment on this PR.

However, with your solution of adding the shared interrupt enabled if both fdcan are enabled as well works great !

One last concerns about shared interrupt was this [statement](https://github.com/zephyrproject-rtos/zephyr/blob/e7875de3f9ad434e1c6e21416c58afcb21ed3acf/doc/kernel/services/interrupts.rst?plain=1#L247)

> Please note that enabling [CONFIG_SHARED_INTERRUPTS](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_SHARED_INTERRUPTS) will result in a non-negligible increase in the binary size. Use with caution.

I have very quickly checked with the counter example:

```
west build -b nucleo_g0b1re zephyr/samples/drivers/can/counter/
```

fdcan2 disabled -> no shared irq -> 

> Memory region         Used Size  Region Size  %age Used
>            FLASH:       57012 B       512 KB     10.87%
>              RAM:       11928 B       144 KB      8.09%

fdcan1 enabled -> shared irq enabled ->

> Memory region         Used Size  Region Size  %age Used
>            FLASH:       57944 B       512 KB     11.05%
>              RAM:       12376 B       144 KB      8.39%

I think this is more than acceptable so I've enabled the fdcan2 by default and updated the doc accordingly.

henrikbrixandersen commented:
Okay, sounds good. Please split this into two commits, though. One for the SoC+DTS, one for the board.

adri1mart1 commented:
Done on my side

erwango reviewed (changes requested):
Please update nucleo_g0b1re.yaml file to mention it supports can.
Aim is that it could be automatically compiled/tested in CI.
Otherwise LGTM, thanks for this contribution

erwango commented on soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx:
```diff
@@ -11,4 +11,7 @@ config SOC
 
+config SHARED_INTERRUPTS
+	default y if $(dt_nodelabel_enabled,fdcan1) && $(dt_nodelabel_enabled,fdcan2)
+
```
This should also depends on CAN. Sorry for the late finding.

adri1mart1 replied:
So just changing it to:

```
config SHARED_INTERRUPTS
        select CAN
	default y if $(dt_nodelabel_enabled,fdcan1) && $(dt_nodelabel_enabled,fdcan2)
```

 ?

henrikbrixandersen commented on soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx:
```diff
@@ -11,4 +11,8 @@ config SOC
 config NUM_IRQS
 	default 31
 
+config SHARED_INTERRUPTS
+	depends on CAN
+	default y if $(dt_nodelabel_enabled,fdcan1) && $(dt_nodelabel_enabled,fdcan2)
+
```
```suggestion

if CAN_STM32_FDCAN

config SHARED_INTERRUPTS
	default y if $(dt_nodelabel_enabled,fdcan1) && $(dt_nodelabel_enabled,fdcan2)

endif # CAN_STM32_FDCAN
```

adri1mart1 replied:
Done 👍

erwango replied:
@henrikbrixandersen In General, recommendation used to be using `depends on FOO` when only one symbol is impacted and `if FOO` when more than one are impacted by the condition.

Also, in general I prefer to keep `BAR_STM32` symbols as much possible within the BAR driver. This is strictly equivalent and quicker to understand:
-`CAN` is well.. `CAN`
-`CAN_STM32_FDCAN` can suggest this is not exactly `CAN`, but some other conditions come into play ? So one would need to check

adri1mart1 replied:
From what I understood and what I've just tested right now with an example:

With the current code, If I compile any program that does not use CAN, I have the SHARED_INTERRUPT enabled (we don't want that) so the condition if indeed needed.

When CONFIG_CAN=y, the flag CONFIG_STM32_FDCAN is by default set as well only if you have at least one controller enabled in your dts.

In the end, both solution works but:
 - Using a dependency with CAN is cleaner but only rely on the fact CONFIG_STM32_FDCAN is enabled by default when CAN is enabled.
 - Using a dependency with CONFIG_STM32_FDCAN is more accurate as right now, the shared IRQ is only for the STM32_FDCAN driver. 

I leave it up to you guys

henrikbrixandersen replied:
> @henrikbrixandersen In General, recommendation used to be using `depends on FOO` when only one symbol is impacted and `if FOO` when more than one are impacted by the condition.

That is correct, but we are not defining a config option here, we are modifying one. We don't want to modify its dependencies to include `CONFIG_CAN` (users should be able to turn this on support for shared IRQs even if not using CAN).

> Also, in general I prefer to keep `BAR_STM32` symbols as much possible within the BAR driver. This is strictly equivalent and quicker to understand: -`CAN` is well.. `CAN` -`CAN_STM32_FDCAN` can suggest this is not exactly `CAN`, but some other conditions come into play ? So one would need to check

Well, you could have `CONFIG_CAN=y` but `CONFIG_CAN_STM32_FDCAN=n` if using an external CAN controller driver only, which again means that there is no need for changing the default of `CONFIG_SHARED_INTERRUPTS`.

henrikbrixandersen replied:
> * Using a dependency with CAN is cleaner but only rely on the fact CONFIG_STM32_FDCAN is enabled by default when CAN is enabled.

... and `CONFIG_STM32_FDCAN` can be disabled by the user.

> * Using a dependency with CONFIG_STM32_FDCAN is more accurate as right now, the shared IRQ is only for the STM32_FDCAN driver.

Exactly.

adri1mart1 commented on soc/arm/st_stm32/stm32g0/Kconfig.defconfig.stm32g0b1xx:
```diff
@@ -11,4 +11,8 @@ config SOC
 	default 31
 
+config SHARED_INTERRUPTS
+	depends on CAN
```
Makes sense, I've just pushed the modification

Commit:
boards: arm: nucleo_g0b1re: add fdcan2 support

The nucleo_g0b1re board is based on stm32g0 that supports 2 can
controllers. This commit adds support for the second can controller.
Add also can label in yml board file.

Signed-off-by: Adrien MARTIN <adrienmar@kickmaker.net>

erwango reviewed (approved):
Approved as I don't want to be picky, but still advise use of `depends on` FOO when only one symbol is impacted (while `if FOO` should be reserved when more than one are impacted by the condition).
